File: hostel_management_system_backend/Controllers/RoomsController.cs
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public sealed class RoomsController : ControllerBase
{
    private readonly IRoomsService _service;

    public RoomsController(IRoomsService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<ActionResult<List<RoomReadDto>>> GetAll(CancellationToken cancellationToken)
    {
        var rooms = await _service.GetAllAsync(cancellationToken);
        return Ok(rooms);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<RoomReadDto>> GetById(Guid id, CancellationToken cancellationToken)
    {
        var room = await _service.GetByIdAsync(id, cancellationToken);
        return room is null ? NotFound() : Ok(room);
    }

    [HttpPost]
    public async Task<ActionResult<RoomReadDto>> Create([FromBody] RoomCreateDto dto, CancellationToken cancellationToken)
    {
        var created = await _service.CreateAsync(dto, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<RoomReadDto>> Update(Guid id, [FromBody] RoomUpdateDto dto, CancellationToken cancellationToken)
    {
        var updated = await _service.UpdateAsync(id, dto, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken cancellationToken)
    {
        var deleted = await _service.DeleteAsync(id, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: src/rooms/rooms.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';

import { RoomCreateDto, RoomReadDto, RoomUpdateDto } from './dto/rooms.dto';
import { RoomsService } from './rooms.service';

@Controller('rooms')
export class RoomsController {
  constructor(private readonly roomsService: RoomsService) {}

  @Get()
  getAll(): Promise<RoomReadDto[]> {
    return this.roomsService.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseUUIDPipe) id: string): Promise<RoomReadDto> {
    return this.roomsService.getById(id);
  }

  @Post()
  create(@Body() dto: RoomCreateDto): Promise<RoomReadDto> {
    return this.roomsService.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: RoomUpdateDto,
  ): Promise<RoomReadDto> {
    return this.roomsService.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseUUIDPipe) id: string) {
    await this.roomsService.delete(id);
  }
}
